cpp/src/join/hash_join.cu

```diff
@@ -151,40 +107,50 @@ get_left_join_indices_complement(rmm::device_vector<size_type> &right_indices,
   // produce exactly the same result as the else path but will be faster.
   if (left_table_row_count == 0) {
     thrust::sequence(rmm::exec_policy(stream),
-                     right_indices_complement.begin(),
-                     right_indices_complement.end(),
+                     right_indices_complement->begin(),
+                     right_indices_complement->end(),
                      0);
   } else {
     // Assume all the indices in invalid_index_map are invalid
-    rmm::device_vector<size_type> invalid_index_map(right_table_row_count, 1);
+    auto invalid_index_map =
+      std::make_unique<rmm::device_uvector<size_type>>(right_table_row_count, stream);
```

Comment:
Why does this need to be a unique pointer rather than just a stack-allocated device_uvector? e.g. 
```suggestion
      rmm::device_uvector<size_type>(right_table_row_count, stream);
```

Reply:
I mentioned this in the PR description: it's because of a Cython limitation. The limitation is that we can't wrap C++ functions for which both the criteria below are true:

1. The C++ function can raise exceptions
2. The function has a return type that doesn't define the default constructor (`rmm::device_uvector` does not (and should not?) have a default ctor)

Reply:
Got it.